server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -178,34 +183,86 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
                         }
                     }
                     if (indexMetaData != null) {
-                        // Find the default pipeline if one is defined from and existing index.
-                        String defaultPipeline = IndexSettings.DEFAULT_PIPELINE.get(indexMetaData.getSettings());
-                        indexRequest.setPipeline(defaultPipeline);
-                        if (IngestService.NOOP_PIPELINE_NAME.equals(defaultPipeline) == false) {
-                            hasIndexRequestsWithPipelines = true;
+                        final Settings indexSettings = indexMetaData.getSettings();
+                        if (IndexSettings.REQUIRED_PIPELINE.exists(indexSettings)) {
+                            // find the required pipeline if one is defined from an existing index
+                            requiredPipeline = IndexSettings.REQUIRED_PIPELINE.get(indexSettings);
+                            assert IndexSettings.DEFAULT_PIPELINE.get(indexSettings).equals(IngestService.NOOP_PIPELINE_NAME) :
+                                IndexSettings.DEFAULT_PIPELINE.get(indexSettings);
+                            indexRequest.setPipeline(requiredPipeline);
+                            requestCanOverridePipeline = false;
+                        } else {
+                            // find the default pipeline if one is defined from an existing index
+                            String defaultPipeline = IndexSettings.DEFAULT_PIPELINE.get(indexSettings);
+                            indexRequest.setPipeline(defaultPipeline);
                         }
                     } else if (indexRequest.index() != null) {
-                        // No index exists yet (and is valid request), so matching index templates to look for a default pipeline
+                        // the index does not exist yet (and is valid request), so match index templates to look for a default pipeline
                         List<IndexTemplateMetaData> templates = MetaDataIndexTemplateService.findTemplates(metaData, indexRequest.index());
                         assert (templates != null);
-                        String defaultPipeline = IngestService.NOOP_PIPELINE_NAME;
-                        // order of templates are highest order first, break if we find a default_pipeline
+                        // order of templates are highest order first, we have to iterate through them all though
+                        String defaultPipeline = null;
                         for (IndexTemplateMetaData template : templates) {
                             final Settings settings = template.settings();
-                            if (IndexSettings.DEFAULT_PIPELINE.exists(settings)) {
+                            if (requiredPipeline == null && IndexSettings.REQUIRED_PIPELINE.exists(settings)) {
```

Comment:
So, if you have 2 (or more) templates that match and 1 has `default_pipeline` and another `required_pipeline` the index creation will fail 
```
required pipeline [test-pipeline] and default pipeline [alt-pipeline] can not both be set
```
I think I understand why we do this (merging of setting would result in this situation). However, I was a bit surprised that `required_pipeline` didn't take precedence. 

Nothing to do here, just wanted to raise the question if we should allow highest order required to take precedence in the case of multi-template matches ?

Reply:
In general I'm against mixing settings / behaviour if two or more index templates happen to match. This also applies in this case, so I favour failing over ignoring the default pipeline.

Reply:
Yeah, this is deliberate. On a given index, I think it's clear we should disallow required and default pipelines, even if they are the same. That's the kind of leniency that we have gotten away from over the years that I think ends up making for a better user experience. Otherwise, if we allowed both, how is a user to reason about which one takes precedence, or if they are both applied somehow, or what? The only way they would know is to consult documentation, and all else equal, the more times a user has to consult the documentation, the worse the experience.

Then it comes to templates, and what should we do? I think we should treat it the same, and we shouldn't break the model of how templates work in general. We take the setting from the highest order template, and there shouldn't be an exception with respect to mixing of settings. Again, it becomes too hard to reason about. As a user, they would think, in general if I have two settings in multiple templates the highest-order one takes effect, and I'm not allowed to have required and default pipelines, unless they come from templates of different orders. Gross.

And now we come to the complexity of template inheritance anyway, something we talked a year ago about getting rid of. It's too complex, with little benefit. It's on me that we haven't taken it further to get rid of them, maybe I'll pick this up again soon.